ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -675,40 +609,93 @@ class ( HasHeader blk
   Internal auxiliary
 -------------------------------------------------------------------------------}
 
-data ValidationResult blk
-  = -- | The chain was valid, the ledger corresponds to the tip of the chain.
-    ValidChain (Chain blk) (ExtLedgerState blk)
-  | InvalidChain
-      (ExtValidationError blk)
-      -- ^ The validation error of the invalid block.
-      (NonEmpty (RealPoint blk))
-      -- ^ The point corresponding to the invalid block is the first in this
-      -- list. The remaining elements in the list are the points after the
-      -- invalid block.
-      (Chain blk)
-      -- ^ The valid prefix of the chain.
-      (ExtLedgerState blk)
-      -- ^ The ledger state corresponding to the tip of the valid prefix of
-      -- the chain.
+type InvalidBlocks blk = Map (HeaderHash blk) (InvalidBlockReason blk, SlotNo)
+
+-- | Result of 'validate', also used internally.
+data ValidationResult blk =
+    ValidationResult
+      !(Chain blk)           -- ^ Valid prefix
+      !(ExtLedgerState blk)  -- ^ Corresponds to the tip of the valid prefix
+      !(InvalidBlocks blk)   -- ^ Invalid blocks encountered while validating
+                             -- the candidate chain.
 
+-- | Validate the given 'Chain'.
+--
+-- The 'InvalidBlocks' in the returned 'ValidationResult' will be >= the
+-- 'invalid' of the given 'Model'.
 validate :: forall blk. LedgerSupportsProtocol blk
          => TopLevelConfig blk
-         -> ExtLedgerState blk
+         -> Model blk
          -> Chain blk
          -> ValidationResult blk
-validate cfg initLedger chain =
-    go initLedger Genesis (Chain.toOldestFirst chain)
+validate cfg Model { currentSlot, maxClockSkew, initLedger, invalid } chain =
+    go initLedger Genesis (Chain.toOldestFirst chain) Nothing
   where
     go :: ExtLedgerState blk  -- ^ Corresponds to the tip of the valid prefix
        -> Chain blk           -- ^ Valid prefix
        -> [blk]               -- ^ Remaining blocks to validate
+       -> Maybe (ValidationResult blk)
+          -- ^ If 'Nothing', continue as normal. If 'Just', continue
+          -- validating, but return the chain and ledger from this
+          -- 'ValidationResult', augmented with any new invalid blocks.
+          --
+          -- This is used to continue validating blocks, even when they're
+          -- from the future, as the real implementation would also validate
+          -- them. This makes sure we detect the same invalid blocks in the
+          -- future.
```

Comment:
Clarify with A (valid) -> B (future) -> C (future) -> D (invalid according the ledger)

The implementation would find B, C as in the future, but not exceeding clock skew, and D as invalid.

We want to start after A, but we have to keep validating to see D, passing around the updated ledger.